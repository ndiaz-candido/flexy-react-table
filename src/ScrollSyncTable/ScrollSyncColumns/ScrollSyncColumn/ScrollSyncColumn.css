.scrollSyncColumn {
  background: #FFF;
  border-right: 1px solid gray;
  padding: 10px;
  width: 120px;
  height: 200px;
}
