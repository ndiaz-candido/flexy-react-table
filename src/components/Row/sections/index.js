import LeftSection from './LeftSection';
import RightSection from './RightSection';
import ScrollableSection from './ScrollableSection';

export { LeftSection, RightSection, ScrollableSection };
