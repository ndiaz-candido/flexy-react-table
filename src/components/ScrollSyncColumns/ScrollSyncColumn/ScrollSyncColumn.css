.scrollSyncColumn {
  padding: 10px;
}
