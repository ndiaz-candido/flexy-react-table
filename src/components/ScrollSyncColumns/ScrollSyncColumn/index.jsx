import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { scrollSyncColumn } from './index.module.css';

const ScrollSyncColumn = ({ width, children, className }) => {
  const styles = {
    width,
    flex: `0 0 ${width}`,
  };
  const classNames = classnames(className, scrollSyncColumn);

  return (
    <div className={classNames} style={styles}>
      {children}
    </div>
  );
};

ScrollSyncColumn.propTypes = {
  width: PropTypes.string,
  children: PropTypes.any.isRequired,

  /** The class name to override the default styles */
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
};

ScrollSyncColumn.defaultProps = {
  width: '150px',
  className: ''
};

export default ScrollSyncColumn;
