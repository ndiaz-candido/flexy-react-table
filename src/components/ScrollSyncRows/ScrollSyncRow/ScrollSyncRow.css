.scrollSyncRow {
  display: flex;
  overflow-x: hidden;
  min-height: 50px;
}

.sticky {
  position: sticky;
  top: 0;
  z-index: 1000;
}
