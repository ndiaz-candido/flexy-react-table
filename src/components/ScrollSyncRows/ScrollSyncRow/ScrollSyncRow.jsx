import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  LeftStickySection,
  RightStickySection,
  ScrollableSection,
} from './sections';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};
const cellDefaultStyles = width => ({
  width,
  flex: `0 0 ${width}`,
});

class ScrollSyncRow extends PureComponent {
  static DEFAULT_COLUMN_WIDTH = 50;
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  rowId = -1;

  constructor(props) {
    super(props);

    this.rowId = Number(props.rowId);
  }

  state = {
    leftStickySection: [],
    rightStickySection: [],
    scrollableSection: [],
  };

  static getDerivedStateFromProps({
    columns,
    children = [],
    columnClassName,
    isHeader,
  }) {
    const leftStickySection = [];
    const rightStickySection = [];
    const scrollableSection = [];

    columns.forEach(column => {
      const {
        dataKey = '',
        width = ScrollSyncRow.DEFAULT_COLUMN_WIDTH,
        stickyAlign,
      } = column.props;
      const cellContent = isHeader
        ? column.props.children
        : React.Children.toArray(children).find(
            cell => cell.props.column === dataKey
          );
      const styles = cellDefaultStyles(width);
      const cell = (
        <div className={columnClassName} style={styles}>
          {cellContent}
        </div>
      );

      switch (stickyAlign) {
        case 'left':
          leftStickySection.push(cell);
          break;
        case 'right':
          rightStickySection.push(cell);
          break;
        default:
          scrollableSection.push(cell);
          break;
      }
    });

    return {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    };
  }

  render() {
    const { isSticky, scrollLeft, onScroll, className } = this.props;
    const {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    } = this.state;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftStickySection cells={leftStickySection} />
        <ScrollableSection
          rowId={this.rowId}
          cells={scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightStickySection cells={rightStickySection} />
      </div>
    );
  }
}

ScrollSyncRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
};

ScrollSyncRow.defaultProps = {
  className: '',
};

export default ScrollSyncRow;
