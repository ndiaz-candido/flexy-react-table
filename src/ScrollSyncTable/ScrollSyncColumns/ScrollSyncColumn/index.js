import ScrollSyncColumn from './ScrollSyncColumn';

export default ScrollSyncColumn;
