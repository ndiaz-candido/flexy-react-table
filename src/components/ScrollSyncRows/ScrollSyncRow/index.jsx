import React, { PureComponent } from 'react';
import ScrollSyncTableCell from '../../ScrollSyncCell';
import LeftStickySection from './Sections/LeftStickySection';
import RightStickySection from './Sections/RightStickySection';
import ScrollableSection from './Sections/ScrollableSection';
import './ScrollSyncRow.css';

class ScrollSyncRow extends PureComponent {
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  rowColumns = [];
  leftStickySection = [];
  rightStickySection = [];
  scrollableSection = [];
  rowId = -1;

  constructor(props) {
    super(props);

    this.rowId = props.rowId;
  }

  componentWillMount() {
    if (this.rowId === ScrollSyncRow.HEADER_ROW_ID) {
      this.rowColumns = this.props.columns;
      this.distribuiteColumnsPerSection();
    } else {
      if (this.areChildrenSyncCells()) {
        this.rowColumns = this.setChildPerColumn();
        this.distribuiteColumnsPerSection();
      } else {
        console.warn('ScrollSyncRow: all the row children must be instances of ScrollSyncCell');
      }
    }
  }

  areChildrenSyncCells = () => {
    const { children } = this.props;
    const childrenArray = React.Children.toArray(children);

    return childrenArray.some((child) => (child.type === ScrollSyncTableCell));
  };

  replaceColumnChildren = (column, index) => {
    const { children } = this.props;
    const columnName = column.props.name;
    const scrollSyncCells = React.Children.toArray(children);
    const scrollSyncCellMatchingColumn = scrollSyncCells.find((cell) => (cell.props.column === columnName));
    const scrollSyncCellMatchingColumnChildren = scrollSyncCellMatchingColumn ? scrollSyncCellMatchingColumn.props.children : '';

    return React.cloneElement(column, { children: scrollSyncCellMatchingColumnChildren });
  }

  setChildPerColumn() {
    const { children, columns } = this.props;
    let rowColumns = [];

    if (children && React.Children.count(children) > 0) {
      rowColumns = columns.map(this.replaceColumnChildren);
    } else {
      console.warn('ScrollSyncRow: no ScrollSyncCells were specified as row children');
    }

    return rowColumns;
  }

  distribuiteColumnsPerSection() {
    this.rowColumns.forEach((column) => {
      switch (column.props.stickyAlign) {
        case 'left': this.leftStickySection.push(column); break;
        case 'rigth': this.rightStickySection.push(column); break;
        default: this.scrollableSection.push(column); break;
      }
    });
  }

  render() {
    const {
      rowId,
      isSticky,
      rowBeingScrolled,
      scrollLeft,
      onScroll
    } = this.props;

    const rowClasses = isSticky ? 'scrollSyncRow sticky' : 'scrollSyncRow';

    return (
      <div className={rowClasses}>
        <LeftStickySection columns={this.leftStickySection} />
        <ScrollableSection
          rowId={rowId}
          columns={this.scrollableSection}
          rowBeingScrolled={rowBeingScrolled}
          scrollLeft={scrollLeft}
          onScroll={onScroll}
        />
        <RightStickySection columns={this.rightStickySection} />
      </div>
    );
  };
};

export default ScrollSyncRow;
