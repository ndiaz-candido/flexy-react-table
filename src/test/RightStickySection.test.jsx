import React from 'react';
import { shallow } from 'enzyme';
import { RightStickySection } from '../components/ScrollSyncRows/ScrollSyncRow/sections';
import ScrollSyncColumn from '../components/ScrollSyncColumns/ScrollSyncColumn';

describe('RightStickySection', () => {
  test('should render its default content', () => {
    const RightStickySectionComponent = shallow(
      <RightStickySection cells={[]} />
    );

    expect(RightStickySectionComponent).toMatchSnapshot();
  });

  test('should render the given columns', () => {
    const columns = [
      <ScrollSyncColumn key="a" dataKey="a">
        Column A
      </ScrollSyncColumn>,
      <ScrollSyncColumn key="b" dataKey="b">
        Column B
      </ScrollSyncColumn>,
    ];
    const RightStickySectionComponent = shallow(
      <RightStickySection cells={columns} />
    );

    expect(RightStickySectionComponent).toMatchSnapshot();
  });
});
