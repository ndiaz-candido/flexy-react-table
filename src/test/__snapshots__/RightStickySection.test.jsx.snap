// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RightStickySection should render its default content 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flex": "0 0 auto",
      "position": "sticky",
      "right": "0",
    }
  }
/>
`;

exports[`RightStickySection should render the given columns 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flex": "0 0 auto",
      "position": "sticky",
      "right": "0",
    }
  }
>
  <React.Fragment
    key="rightSection=0"
  >
    <ScrollSyncColumn
      className=""
      dataKey="a"
      key="a"
      width="150px"
    >
      Column A
    </ScrollSyncColumn>
  </React.Fragment>
  <React.Fragment
    key="rightSection=1"
  >
    <ScrollSyncColumn
      className=""
      dataKey="b"
      key="b"
      width="150px"
    >
      Column B
    </ScrollSyncColumn>
  </React.Fragment>
</div>
`;
